### PR TITLE
Fix with_retries

### PR DESCRIPTION
**Pull request by joshlory**
Before:

``` ruby
irb(main):009:0> with_retries(3) { puts 'try'; raise 'fail' }
try
=> 3
```

After:

``` ruby
irb(main):008:0> with_retries(3) { puts 'try'; raise 'fail' }
try
try
try
RuntimeError: fail
        from (irb):10:in `block in irb_binding'
        from (irb):2:in `with_retries'
        from (irb):8
        from /Users/joshlory/.rbenv/versions/2.0.0-p481/bin/irb:12:in `<main>'
```

This was caught by `rubocop --only Lint/UselessAssignment`, which warned on the unused `count` variable.

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -17,13 +17,12 @@ def create_database(uri)
...
-    raise if (count -= 1) == 0
-    sleep 2.25
-  end
+def with_retries(count = 5)
+  yield if block_given?
+rescue
+  raise if (count -= 1) == 0
+  sleep 2.25
+  retry
 end
```
Comment (wjordan) [inline Rakefile]: I'd never seen this syntax before, so I had to [look it up](https://github.com/ruby/ruby/blob/trunk/doc/syntax/exceptions.rdoc) to understand the behavior. Two things I learned:

1.
>If you are inside a method you do not need to use begin or end unless you wish to limit the scope of rescued exceptions:
```
def my_method
  # ...
rescue
  # ...
end
```

2.
> You may retry rescued exceptions:
```
begin
  # ...
rescue
  # do something that may change the result of the begin block
  retry
end
```
Execution will resume at the start of the begin block, so be careful not to create an infinite loop.

**Comment by wjordan** [comment]
LGTM. Learned some new syntax!

File: Rakefile (hunk 1 above)
Comment (laurelfan) [inline Rakefile]: in dashboard we also use retryable gem, which is essentially with_retries with a couple more options (in particular, filtering on the exception you want to retry)

Comment (joshlory) [inline Rakefile]: Makes sense to use this gem if we're already using it elsewhere.  Fixed in ade19e3.

File: pegasus/helpers/section_api_helpers.rb [hunk 2]
```diff
@@ -216,8 +216,7 @@ def self.create(params)
     created_at = DateTime.now
 
     row = nil
-    tries = 0
-    begin
+    retryable(tries: 2, sleep: 0, on: Sequel::UniqueConstraintViolation) do
```
Comment (laurelfan) [inline pegasus/helpers/section_api_helpers.rb]: is retryable already `require`d in pegasus?

File: pegasus/helpers/hoc_helpers.rb [hunk 3]
```diff
@@ -2,15 +2,10 @@
 require 'cdo/graphics/certificate_image'
 
 def create_session_row(row)
```
Comment (laurelfan) [inline pegasus/helpers/hoc_helpers.rb]: to test, hit the tracking pixel at http://code.org/api/hour/begin_codeorg.png (sub localhost/staging/etc)

File: pegasus/helpers/section_api_helpers.rb (hunk 2 above)
Comment (laurelfan) [inline pegasus/helpers/section_api_helpers.rb]: to test, go to teacher dashboard at http://code.org/teacher-dashboard and create a section

**Comment by joshlory** [comment]
Backing out `retryable` gem changes for now -- they are in `branch:fix-with-retries-next`.

**Pull request merged**